lib/perl/Genome/Model/Build/ReferenceSequence/IndexBase.pm

```diff
@@ -0,0 +1,277 @@
+        $self->error_message("Failed to create AlignmentIndex objects for dependencies");
+        return;
+    }
+
+    return $self;
+}
+
+sub _resolve_reference_fasta_file {
+    my $self = shift;
+
+    if ($self->_supports_multiple_reference) {
+        return $self->reference_build->primary_consensus_path('fa');
+    } else {
+        return $self->reference_build->full_consensus_path('fa');
+    }
+}
+
+sub _prepare_index {
+    my $self = shift;
+
+    Carp::confess('Class must implement _prepare_index.');
+}
+
+sub _promote_prepared_index {
+    my $self = shift;
+
+    my $output_dir = $self->output_dir || $self->_prepare_output_directory;
+    $self->debug_message("Alignment output path is $output_dir");
+
+    unless ($self->_promote_data)  {
+        $self->error_message("Failed to de-stage data into output path " . $self->output_dir);
+        return;
+    }
+
+    $self->_reallocate_disk_allocation;
+
+    $self->debug_message("Prepared alignment reference index!");
+
+    return 1;
+}
+
+sub aligner_requires_param_masking {
+    my $class = shift;
+    my $aligner_name = shift;
+
+    # If $aligner_name is not known then we can't ask.  While this could be an
+    # exception it is the case that while an object is being created its
+    # params/inputs getted added one-by-one which means sometime $aligner_name
+    # was not yet known.
+    unless ($aligner_name) {
+        return 0;
+    }
+
+    my $aligner_class = 'Genome::InstrumentData::AlignmentResult::'  . Genome::InstrumentData::AlignmentResult->_resolve_subclass_name_for_aligner_name($aligner_name);
+
+    # if aligner params are not required for index, and we can   generically create an index for that version, then filter it out.
+    if ($aligner_class->aligner_params_required_for_index) {
+        $class->debug_message("This aligner requires a parameter-specific index.  Can't mask params out.");
+        return 0;
+    }
+
+    return 1;
+}
+
+sub full_consensus_path {
+    my $self = shift;
+    my $extension = shift;
+
+    my $path = $self->output_dir . '/all_sequences';
+    if ($extension) {
+        $path .= '.'. $extension;
+    }
+    return $path;
+}
+
+sub get_or_create {
+    my $class = shift;
+    my %params = @_;
+
+    my @objects = $class->SUPER::get_or_create(%params);
+
+    for my $obj (@objects) {
+        next unless ref($obj); # sometimes UR gives us back the package name when deleting?
+        unless ($obj->generate_dependencies_as_needed($params{users})) {
+            $obj->error_message("Failed to get AlignmentIndex objects for dependencies of " . $obj->__display_name__);
+            return;
+        }
+    }
+
+    if (@objects > 1) {
+        return @objects if wantarray;
+        my @ids = map { $_->id } @objects;
+        die "Multiple matches for $class but get or create was called in scalar context! Found ids: @ids";
+    }
+    else {
+        return $objects[0];
+    }
+}
+
+sub _gather_params_for_get_or_create {
```

Comment:
This wasn't in AnnotationIndex before, but it seems ok for it to be in there now (just checking again!)